File: src/main/java/com/gitlab/daring/image/command/parameter/ParamFileChooser.java
package com.gitlab.daring.image.command.parameter;

import one.util.streamex.StreamEx;

import javax.swing.*;
import javax.swing.filechooser.FileFilter;
import javax.swing.filechooser.FileNameExtensionFilter;
import java.io.File;
import java.nio.file.Path;
import java.nio.file.Paths;

import static com.gitlab.daring.image.swing.SwingUtils.newButton;
import static javax.swing.JFileChooser.APPROVE_OPTION;

class ParamFileChooser {

    public final FileParam param;
    public final JTextField valueField;
    public final JButton openButton;
    
    JFileChooser fileChooser;

    public ParamFileChooser(FileParam param, JTextField valueField) {
        this.param = param;
        this.valueField = valueField;
        openButton = newButton("...", this::open);
    }

    JFileChooser createFileChooser() {
        JFileChooser fc = new JFileChooser();
        fc.setCurrentDirectory(new File("."));
        fc.setMultiSelectionEnabled(param.maxCount != 1);
        fc.setFileFilter(createFileFilter());
        return fc;
    }

    FileFilter createFileFilter() {
        String[] exts = {"bmp", "png", "jpg", "jpeg", "gif"};
        return new FileNameExtensionFilter("Images", exts);
    }

    public void open() {
        if (fileChooser == null) fileChooser = createFileChooser();
        int r = fileChooser.showOpenDialog(valueField);
        if (r == APPROVE_OPTION) {
            String fl = buildFileList();
            valueField.setText(fl);
            valueField.postActionEvent();
        }
    }

    String buildFileList() {
        Path d = Paths.get("").toAbsolutePath();
        File[] fs = fileChooser.getSelectedFiles();
        return StreamEx.of(fs).map(File::toPath).map(d::relativize).joining(",");
    }

}

File: src/main/kotlin/com/gitlab/daring/image/command/parameter/ParamFileChooser.kt
package com.gitlab.daring.image.command.parameter

import com.gitlab.daring.image.swing.SwingUtils.newButton
import org.apache.commons.lang3.StringUtils.defaultIfBlank
import java.io.File
import java.nio.file.Paths
import javax.swing.JButton
import javax.swing.JFileChooser
import javax.swing.JFileChooser.APPROVE_OPTION
import javax.swing.JTextField
import javax.swing.filechooser.FileFilter
import javax.swing.filechooser.FileNameExtensionFilter

class ParamFileChooser(val param: FileParam, val valueField: JTextField) {

    val openButton: JButton = newButton("...", this::open)

    val fc by lazy { createFileChooser() }

    fun createFileChooser(): JFileChooser {
        val fc = JFileChooser()
        val vf = File(defaultIfBlank(param.v, "."))
        fc.currentDirectory = if (vf.isFile) vf.parentFile else vf
        fc.isMultiSelectionEnabled = param.maxCount != 1
        fc.fileFilter = createFileFilter()
        return fc
    }

    fun createFileFilter(): FileFilter {
        val exts = arrayOf("bmp", "png", "jpg", "jpeg", "gif")
        return FileNameExtensionFilter("Images", *exts)
    }

    fun open() {
        val r = fc.showOpenDialog(valueField)
        if (r == APPROVE_OPTION) {
            valueField.text = buildFileList()
            valueField.postActionEvent()
        }
    }

    fun buildFileList(): String {
        val d = Paths.get("").toAbsolutePath()
        val fs = if (fc.isMultiSelectionEnabled) fc.selectedFiles else arrayOf(fc.selectedFile)
        return fs.map(File::toPath).map { p ->
            try { d.relativize(p) } catch (e: Exception) { p }
        }.joinToString(",");
    }

}